### PR TITLE
BAH-1397 | Introduction of additional identifier lookup with generic extension point with specified input fields validation

### PR DESCRIPTION
**Pull request by SanoferSameera**

### CODE REVIEW

File: ui/app/registration/controllers/patientCommonController.js [hunk 1]
```diff
@@ -17,7 +17,186 @@ angular.module('bahmni.registration')
...
+                            if (identifierType === identifierTypes[j]) {
+                                return $scope.regExtPoints[i];
+                            }
+                        }
+                    }
+                }
+                return null;
+            }
+
+            function update (patient) {
```
Comment (angshu) [inline ui/app/registration/controllers/patientCommonController.js:94]: Can we put better name for this function? if this is about updating information from other sources - say something like updateInfoFromExtSource()

File: ui/app/registration/controllers/patientCommonController.js [hunk 2]
```diff
@@ -17,7 +17,186 @@ angular.module('bahmni.registration')
...
+                        break;
+                    }
+                }
+            }
+
+            $scope.closeIdentifierPopup = function () {
+                $scope.showIframe = false;
+            };
+
+            function calculateAge (birthDate) {
```
Comment (angshu) [inline ui/app/registration/controllers/patientCommonController.js:166]: there should be already a function that calculates patient age from a date. Check createPatientContrroller or editPatientController .. you can pull that out to a common utils

File: ui/app/registration/controllers/patientCommonController.js [hunk 3]
```diff
@@ -190,6 +369,25 @@ angular.module('bahmni.registration')
             $scope.$watch('patientLoaded', function () {
                 if ($scope.patientLoaded) {
                     executeShowOrHideRules();
+                    if ($scope.patient.extraIdentifiers !== undefined && $scope.idStatus === false) {
+                        for (var i = 0; i < $scope.patient.extraIdentifiers.length; i++) {
```
Comment (angshu) [inline ui/app/registration/controllers/patientCommonController.js:373]: You can replace with array.forEach() .. also maybe better to extract to a function?

File: ui/app/registration/views/topDownAddressFields.html [hunk 4]
```diff
@@ -11,7 +11,7 @@
                    pattern-validate
                    non-blank="::addressLevel.required"
                    ng-model="address[addressLevel.addressField]"
-                   ng-disabled="isReadOnly(addressLevel)"
+                   ng-disabled="isReadOnly(addressLevel) || isDisabled('address1')"
```
Comment (angshu) [inline ui/app/registration/views/topDownAddressFields.html:14]: why this hardcoding? also no gaurantee that address1 field is mapped, right?

File: ui/test/unit/registration/controllers/editPatientController.spec.js [hunk 5]
```diff
@@ -38,6 +38,14 @@ describe('EditPatientController', function () {
                 }
             }
         };
+
+        patientServiceMock.getAllPatientIdentifiers = function (uuid) {
```
Comment (angshu) [inline ui/test/unit/registration/controllers/editPatientController.spec.js:42]: please add a test with sample patient data (with address)

**Review by angshu — CHANGES_REQUESTED** [review]

File: ui/app/common/patient/services/patientService.js [hunk 6]
```diff
@@ -29,15 +29,25 @@ angular.module('bahmni.common.patient')
...
         };
+        var isHealthIdentifier = function (query) {
+            return !isNaN(query) || query.includes("@");
+        };
 
         this.search = function (query, offset, identifier) {
             offset = offset || 0;
-            return $http.get(Bahmni.Common.Constants.bahmniSearchUrl + "/patient", {
+            var url = "/patient";
+            if (isHealthIdentifier(query)) {
```
[the remark this answers is not in the document]
Comment (SanoferSameera) [reply ui/app/common/patient/services/patientService.js:39]: In clinics, while searching for a patient with extra Identifiers, we need lucene search.

File: ui/app/registration/controllers/patientCommonController.js (hunk 2 above)
Comment (SanoferSameera) [reply ui/app/registration/controllers/patientCommonController.js]: reused the existing function

File: ui/app/common/attributeTypes/views/attributeInformation.html [hunk 7]
```diff
@@ -35,8 +35,9 @@
     <div class="left" data-ng-switch-when="org.openmrs.customdatatype.datatype.FreeTextDatatype">
         <input
                 type="text" pattern-validate id="{{::attribute.name}}"
-                ng-model="targetModel[attribute.name]" data-item-type="personAttribute"  pattern="^\S.*$"
-                ng-disabled="::isReadOnly(attribute.name)" ng-required="{{::attribute.required}}" ng-blur="::handleUpdate(attribute.name)"/>
+                ng-model="targetModel[attribute.name]" data-item-type="personAttribute"
+                ng-disabled="::isReadOnly(attribute.name)" ng-required="{{::attribute.required}}" ng-blur="::handleUpdate(attribute.name)"
+                pattern="[ a-zA-Z]{0,}"/>
```
[the remark this answers is not in the document]
Comment (SanoferSameera) [reply ui/app/common/attributeTypes/views/attributeInformation.html]: removed all the pattern validation

File: ui/app/registration/controllers/patientCommonController.js (hunk 1 above)
Comment (SanoferSameera) [reply ui/app/registration/controllers/patientCommonController.js]: updated the function name

File: ui/app/common/patient/services/patientService.js [hunk 8]
```diff
@@ -32,12 +32,17 @@ angular.module('bahmni.common.patient')
 
         this.search = function (query, offset, identifier) {
             offset = offset || 0;
-            return $http.get(Bahmni.Common.Constants.bahmniSearchUrl + "/patient", {
+            var url = "/patient";
+            identifier = query;
+            url = url + "/lucene";
+            return $http.get(Bahmni.Common.Constants.bahmniSearchUrl + url, {
```
Comment (angshu) [inline ui/app/common/patient/services/patientService.js:38]: Why not create another constant for the URL? "/patinet/lucene"

File: ui/app/registration/controllers/patientCommonController.js [hunk 9]
```diff
@@ -17,7 +17,153 @@ angular.module('bahmni.registration')
...
+                                $window.open(Bahmni.Registration.Constants.existingPatient + $rootScope.extenstionPatient.id, "_self");
+                            }
+                        } else $window.open(Bahmni.Registration.Constants.newPatient, "_self");
+                        $scope.updateInfoFromExtSource($rootScope.extenstionPatient);
+                        $scope.$digest();
+                    }
+                }, false);
+            };
+
+            $scope.isDisabled = function (attribute) {
```
Comment (angshu) [inline ui/app/registration/controllers/patientCommonController.js:44]: maybe better to be more specific function name.. like "isDisabledAttribute()"

File: ui/app/registration/controllers/patientCommonController.js [hunk 10]
```diff
@@ -17,7 +17,153 @@ angular.module('bahmni.registration')
...
+                        var identifier = $rootScope.patientIdentifiers[i];
+                        if (identifier.identifierType.display === identifierType) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+
+            $scope.showButton = function (identifierType, identifierValue) {
```
Comment (angshu) [inline ui/app/registration/controllers/patientCommonController.js:60]: more specific name? showIdentifierVerificationButton()

File: ui/app/registration/controllers/patientCommonController.js [hunk 11]
```diff
@@ -17,7 +17,153 @@ angular.module('bahmni.registration')
             $scope.readOnlyExtraIdentifiers = appService.getAppDescriptor().getConfigValue("readOnlyExtraIdentifiers");
             $scope.showSaveConfirmDialogConfig = appService.getAppDescriptor().getConfigValue("showSaveConfirmDialog");
             $scope.showSaveAndContinueButton = false;
-            $scope.moduleName = appService.getAppDescriptor().getConfigValue('registrationModuleName');
+            $scope.regExtPoints = appService.getAppDescriptor().getExtensions("org.bahmni.registration.identifier", "link");
+
+            $scope.showIframe = false;
+            var buttonMap = new Map();
```
Comment (angshu) [inline ui/app/registration/controllers/patientCommonController.js:23]: Can we think of better name? if this is storing identifier verification to extensions - "identifierExtnMap"

**Review by angshu — CHANGES_REQUESTED** [review]

File: ui/app/common/patient/services/patientService.js (hunk 8 above)
Comment (SanoferSameera) [reply ui/app/common/patient/services/patientService.js:37]: modified

File: ui/app/registration/controllers/patientCommonController.js (hunk 9 above)
Comment (SanoferSameera) [reply ui/app/registration/controllers/patientCommonController.js]: changed to better function name

**Review by angshu — APPROVED** [review]

**Pull request merged**